pkg/controller/clusterstorage/clusterstorage_controller.go

```diff
@@ -183,7 +184,47 @@ func (r *ReconcileClusterStorage) Reconcile(request reconcile.Request) (reconcil
 		return reconcile.Result{}, err
 	}
 
-	// StorageClass already exists - don't requeue
+	// Check to see if modifications have been made to the StorageClass attributes
+	needUpdate := false
+	newSCC := sc.DeepCopy()
+	newSCC.ObjectMeta = found.ObjectMeta
+
+	// Define a default ReclaimPolicy for comparison
+	if newSCC.ReclaimPolicy == nil {
+		deletePolicy := corev1.PersistentVolumeReclaimDelete
+		newSCC.ReclaimPolicy = &deletePolicy
+	}
+
+	if !reflect.DeepEqual(newSCC, found) {
+		needUpdate = true
+	}
+
+	// If a change has been detected, delete and recreate the StorageClass, preserving annotations
+	// so that the default StorageClass value is not overridden.
+	if needUpdate {
+		reqLogger.Info("StorageClass already exists and needs to be recreated", "StorageClass.Name", found.Name)
+		annotations := found.Annotations
+		labels := found.Labels
+		err = r.client.Delete(context.TODO(), found)
```

Comment:
If the operator wants to change just the fields that are changeable, the operator should do `Update`.

Re-create is dangerous, because it leaves the cluster without any default storage class for a short while and newly created PVCs will not get it. It will leave the PVCs without any storage class until user notices it and does something about it manually. I would even say that we won't re-create the storage class at all and just report an error.

Reply:
Discussed this in our standup call - it looks like choosing between two bad options. If we don't re-create SC then that would mean we could *never* update things like SC's parameter or binding mode.

On the other hand if we delete and re-create SC then there could be a short period during upgrade process where no SC will be found.

Reply:
After-meeting thoughts:

I talked to @vrutkovs, currently it seems that major update (e.g. 4.2->4.3) will not be automatic. We can expect that admin reads release notes and can see that they should not create PVCs during the update. We're going to modify immutable SC fields only when really really necessary (if ever) and only during such major update. In that case, I would say that documentation is enough ("check your vSphere PVCs after update").

During minor update (or during normal operation):

- If user changes fields of the storage class, the operator should `Update()` it.
- What if user re-creates the storage class (to change immutable field)? We should clearly document that the storage class created by the operator is ours and users should not change it except for the default annotation. But then? Should the operator re-create (with an event? where? it would be great to get a metrics / alert for this). Now that's clear user error I am open to re-create the SC.

Reply:
I don't think exact Cincinnati schema for major updates has been prepared yet (cc @steveeJ @wking), so its unlikely to happen in a full automatic way in any case

Reply:
I'm not aware of any special casing for major updates, so I assume that if the cluster is configured to automatically apply updates a major one will be no different from a minor one. Of course we could attach metadata to any release and teach the the CVO to behave differently when seeing such an update.

Reply:
For now I've updated the Operator so that it will perform an Update if it notices differences between the two StorageClass entities. If it errors attempting to update an invalid field, then the Operator will enter a degraded state and log the error.

Marking this discussion as resolved for now.